### PR TITLE
Add installation and packaging targets [guanpu]

### PR DESCRIPTION
**Pull request by lVentus**
GitLab account: guanpu

Installation

Build the image:
`docker build -t IMAGENAME .`

Run in interactive mode by mounting the current folder (Ensure you are inside the git repo's root directory):
`docker run --rm -it --mount type=bind,source="$(pwd)",target=/mnt/cpack-exercise IMAGENAME`

I have implemented docker automate building process.
You can find the debain package in dir:
`/mnt/cpack-exercise/output`

But to test the build process manuely, follow these steps:
```
cd /mnt/cpack-exercise/
mkdir -p build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make 
make package
apt install ./cpackexample-0.1.0-Linux.deb
/usr/bin/cpackexample
```
Solved bonus exercise tasks
1. File size comparison
The comparison of the tar.gz package with and without file stripping results in:

du -h cpackexample-0.1.0-Linux-unstripped.deb
4.4M	cpackexample-0.1.0-Linux-unstripped.deb
du -h cpackexample-0.1.0-Linux.deb
2.5M	cpackexample-0.1.0-Linux.deb
Unstripped cpackexample package needs1.9M more space.

2.. Fixed lintian warnings 
First output:
```
E: cpackexample: extended-description-is-empty
E: cpackexample: malformed-contact Maintainer Guanpu
E: cpackexample: missing-dependency-on-libc needed by usr/bin/cpackexample
E: cpackexample: no-changelog usr/share/doc/cpackexample/changelog.gz (native package)
E: cpackexample: no-copyright-file
W: cpackexample: no-manual-page [usr/bin/cpackexample]
```

Final output:
```
W: cpackexample: no-manual-page [usr/bin/cpackexample]
```

Error: extended-description-is-empty
Fixed bu adding `set(CPACK_PACKAGE_DESCRIPTION "This is SSE cpack exercise")`

Error: malformed-contact Maintainer Guanpu
Fixed by adding 
`set(CPACK_DEBIAN_PACKAGE_MAINTAINER "Guanpu <st181070@stud.uni-stuttgart.de>")`

Error: missing-dependency-on-libc needed by usr/bin/cpackexample
Fixed by adding libc6 dependency

Error: no-ropyright-file Copyright file missing
Fixed by adding 
`
install(FILES "${CMAKE_CURRENT_SOURCE_DIR}/cmake/debian/copyright" DESTINATION "${CMAKE_INSTALL_DOCDIR}")
`

Error: no-changelog-file Changelog file missing
Fixed by adding 
```
add_custom_command(
    OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/changelog.gz"
    COMMAND gzip -9 --no-name -c "${CMAKE_CURRENT_SOURCE_DIR}/cmake/debian/changelog" > "${CMAKE_CURRENT_BINARY_DIR}/changelog.gz"
    DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/cmake/debian/changelog"
    COMMENT "Compressing changelog file"
)

add_custom_target(compress_changelog ALL DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/changelog.gz")

install(FILES "${CMAKE_CURRENT_BINARY_DIR}/changelog.gz" DESTINATION "${CMAKE_INSTALL_DOCDIR}")
```
Warning: changelog-not-compressed-with-max-compression
Fixed by using the option -9 for the highest compression with gzip when compressing changelog

Warning: package-contains-timestamped-gzip 2024-12-10T19:13:16 [usr/share/doc/cpackexample/changelog.gz]
Fixed by using the option -no-name

Warning: no-manual-page [usr/bin/cpackexample]
didn't fix

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,19 +4,39 @@ project("cpackexample" VERSION 0.1.0)
...
 
 DEAL_II_SETUP_TARGET("${PROJECT_NAME}")
 DEAL_II_SETUP_TARGET(cpackexamplelib)
+
+include(GNUInstallDirs)
+
+install(TARGETS cpackexamplelib
+        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/cpackexamplelib)
```
Comment (uekerman) [inline CMakeLists.txt:33]:
```suggestion
        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/cpackexample)
```

**Review by uekerman — APPROVED** [review]
Debian package did not have the standard name. Rest looks good :+1: 

The automated package creation did not work for me. How should I run the container exactly?

**Comment by lVentus** [comment]
> Debian package did not have the standard name. Rest looks good 👍
> 
> The automated package creation did not work for me. How should I run the container exactly?

I just build the image with ' docker build -t ex5 .' and use `docker run it ex5`, the result should be in output folder.
image
image

What message did you get when auto package doesn't work?

File: cmake/CPackConfig.cmake [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+set(CPACK_PACKAGE_VENDOR "SSE")
+set(CPACK_PACKAGE_CONTACT "st181070@stud.uni-stuttgart.de")
+set(CPACK_PACKAGE_HOMEPAGE_URL "https://github.com/lVentus/cpack-exercise-wt2425")
+
+set(CPACK_GENERATOR "TGZ;DEB")
+
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "Guanpu <st181070@stud.uni-stuttgart.de>")
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libyaml-cpp0.8, libboost-filesystem-dev, libc6")
+set(CPACK_DEBIAN_PACKAGE_FILE_NAME "${CPACK_PACKAGE_NAME}_${CPACK_PACKAGE_VERSION}-${CMAKE_SYSTEM_PROCESSOR}")
```
[the remark this answers is not in the document]
Comment (lVentus) [reply cmake/CPackConfig.cmake:14]: sorry, I didn't notice much about name

**Comment by uekerman** [comment]
The output needs to be on the host system, not in the container. The container should not run in interactive mode.

**Comment by lVentus** [comment]
> The output needs to be on the host system, not in the container. The container should not run in interactive mode.

Ah sorry, I missunderstand the task. Do I still have chance to change?

**Comment by uekerman** [comment]
I am happy to review any updates. Of course, grading also needs to take the deadline into account.

**Pull request closed without merging**